scripts/deleteSnapsets.js

```diff
@@ -0,0 +1,162 @@
+/*!
+Copyright 2018 OCAD University
+
+Licensed under the New BSD license. You may not use this file except in
+compliance with this License.
+
+You may obtain a copy of the License at
+https://github.com/GPII/universal/blob/master/LICENSE.txt
+*/
+
+// This script modifies the preferences data base:
+// 1. Finds all the Prefs Safes of type "snapset" (prefsSafesType = "snapset"),
+// 2. Finds all the GPII Keys associated with each snapset Prefs Safe
+// 3. Deletes the found Prefs Safes and associated GPII Keys
+//
+// A sample command that runs this script:
+// node deleteSnapsets.js $COUCHDBURL
+
+"use strict";
+
+var http = require("http"),
+    url = require("url"),
+    fluid = require("infusion");
+
+var gpii = fluid.registerNamespace("gpii");
+fluid.registerNamespace("gpii.dataLoader");
+fluid.setLogging(fluid.logLevel.INFO)
+
+var dbLoader = gpii.dataLoader;
+dbLoader.couchDbUrl = process.argv[2];
+if (!fluid.isValue(dbLoader.couchDbUrl)) {
+    fluid.log ("COUCHDB_URL environment variable must be defined");
+    process.exit(1);
+}
+fluid.log("COUCHDB_URL: '" + dbLoader.couchDbUrl + "'");
+dbLoader.prefsSafesViewUrl = dbLoader.couchDbUrl + "/_design/views/_view/findSnapsetPrefsSafes";
+dbLoader.gpiiKeyViewUrl = dbLoader.couchDbUrl + "/_design/views/_view/findGpiiKeysByPrefsSafeId";
+dbLoader.parsedCouchDbUrl = url.parse(dbLoader.couchDbUrl);
+dbLoader.docsToRemove=[];
+
+/**
+ * Find the Prefs Safes of type "snapSet", add them to an array of records to
+ * remove, then use them to find their associated GPII Key(s).
+ * @param {Object} response - The response from the data base containing the
+ *                            snapSet Prefs Safes as a JSON string.
+ */
+dbLoader.processSnapsets = function (response) {
+    var snapSetsString = "";
+    response.setEncoding("utf8");
+    response.on("data", function (chunk) {
+        snapSetsString += chunk;
+    });
+    response.on("end", function () {
+        dbLoader.snapSets = JSON.parse(snapSetsString);
+        fluid.each(dbLoader.snapSets.rows, function (aSnapset) {
+            aSnapset.value._deleted = true;
+            dbLoader.docsToRemove.push(aSnapset.value);
+        });
+        dbLoader.addGpiiKeysAndBulkDelete(dbLoader.snapSets.rows, dbLoader.docsToRemove);
+    });
+};
+
+/**
+ * Find the GPII key records associated with the given snapset Prefs Safes,
+ * delete each one as found, and then batch delete all the snapset Prefs Safes.
+ * @param {Array} snapSets      - The snapsets of interest.
+ * @param {Array} docsToRemove  - Array of records to delete.
+ */
+dbLoader.addGpiiKeysAndBulkDelete = function (snapSets, docsToRemove) {
+    fluid.each(snapSets, function (aSnapset) {
+        var gpiiKeyId = aSnapset.value._id;
+        fluid.log("Snapset: " + gpiiKeyId);
+        var gpiiKeyViewUrl = dbLoader.gpiiKeyViewUrl + "?key=%22" + gpiiKeyId + "%22";
+        var getGpiiKeysRequest = http.request(gpiiKeyViewUrl, function (resp) {
```

Comment:
Worrying lack of abstraction in all of this work. At the least there should be a wrapper for the process of invoking an HTTP request and waiting for responses from it as some kind of promise-producer. I don't want to slow down this work one atom so I'm not suggesting that it be re-expressed using https://github.com/fluid-project/kettle/blob/master/docs/DataSources.md#simple-example-of-using-an-http-datasource but something needs to happen to unwrap the tangle in this method which by line 88 we are 4 closures deep. It's hard to follow this expression of the algorithm, it is brittle, and seems prone to races which are already causing confusion. Also consider utilities like https://docs.fluidproject.org/infusion/development/PromisesAPI.html#fluidpromisesequencesources-options